.github/workflows/dev_version_script.py

```diff
@@ -11,74 +11,102 @@
+                    return semver.Version.parse(parsed_semver)
+                raise ValueError(f"Unable to find valid semver for __version__. Got: '{line}'")
     raise ValueError("Cannot parse version")
 
 
-def is_dev(version_str):
-    m = rgx_dev_ver.fullmatch(version_str)
-    return m is not None
+def update_prerelease_version(repo_root_path: Path, version: semver.Version):
+    """
+    Updates the version file within pennylane_lightning/core/_version.py.
 
+    :param repo_root_path: Path to the repository root.
+    :param version: The new version to use within the file.
+    :return:
+    """
+    version_file_path = repo_root_path / VERSION_FILE_PATH
+    if not version_file_path.exists():
+        raise FileNotFoundError(f"Unable to find version file at location {version_file_path}")
 
-def update_dev_version(package_path, version_str):
-    m = rgx_dev_ver.fullmatch(version_str)
-    if m.group(2) == "":
-        curr_dev_ver = 0
-    else:
-        curr_dev_ver = int(m.group(2))
+    with version_file_path.open() as f:
+        lines = [
+            rgx_ver.sub(f"__version__ = \"{str(version)}\"", line)
+            for line in f
+        ]
 
-    new_version_str = "{}-dev{}".format(m.group(1), str(curr_dev_ver + 1))
-
-    lines = []
-    with package_path.joinpath(VERSION_FILE_PATH).open("r") as f:
-        for line in f.readlines():
-            if not line.startswith("__version__"):
-                lines.append(line)
-            else:
-                lines.append(f'__version__ = "{new_version_str}"\n')
-
-    with package_path.joinpath(VERSION_FILE_PATH).open("w") as f:
+    with version_file_path.open("w") as f:
         f.write("".join(lines))
 
 
 if __name__ == "__main__":
     parser = argparse.ArgumentParser()
-    parser.add_argument("--pr-path", dest="pr", type=str, required=True, help="Path to the PR dir")
+    parser.add_argument("--pr-path", dest="pr", type=Path, required=True, help="Path to the PR dir")
     parser.add_argument(
-        "--master-path", dest="master", type=str, required=True, help="Path to the master dir"
+        "--master-path", dest="master", type=Path, required=True, help="Path to the master dir"
     )
 
     args = parser.parse_args()
 
-    pr_version = extract_version(Path(args.pr))
-    master_version = extract_version(Path(args.master))
+    pr_version = extract_version(args.pr)
+    master_version = extract_version(args.master)
 
-    if pr_version == master_version:
-        if is_dev(pr_version):
-            print("Automatically update version string.")
-            update_dev_version(Path(args.pr), pr_version)
+    print("Got Package Version from 'master' ->", str(master_version))
+    print("Got Package Version from 'pr' ->", str(pr_version))
+
+    # Only attempt to bump the version if the pull_request is:
+    #  - A prerelease, has `X.Y.Z-prerelease` in _version.py
+    #  - The prerelease startswith `dev`. We do not want to auto bump for non-dev prerelease.
```

Comment:
We should likely want this to work for RC too, support both https://peps.python.org/pep-0440/#developmental-releases and https://peps.python.org/pep-0440/#pre-releases. Though, we can add RC in a follow-up if this solves the dev issue for now.

Reply:
Supporting RC would likely be a future feature to add. Even as it stands right now, this update script (what exists on master) branch does not support RC ([See Here](https://github.com/PennyLaneAI/pennylane-lightning/blob/288b08d1f491319fde56307a27d8283b8e8a846b/.github/workflows/dev_version_script.py#L37-L39)).

I retained that logic as part of the refactor, but we can change it in the future :+1: